SourceCode/raytracer/CRTTypes.h

```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include <iostream>
+#include <vector>
+#include <stdexcept>
+#include <cmath>
+
+struct Pixel
+{
+    uint8_t r = 0;
+    uint8_t g = 0;
+    uint8_t b = 0;
+    std::string toString() const
+    {
+        return std::to_string(r) + " " + std::to_string(g) + " " + std::to_string(b) + "\t";
+    }
+};
+
+class Buffer
+{
+public :
+    Buffer(uint16_t _width, uint16_t _height) : width(_width), height(_height)
+    {
+        data = std::make_unique<Pixel[]>(width * height);
+    }
+
+    std::unique_ptr<Pixel[]> data;
+
+    const uint16_t width;
+    const uint16_t height;
+};
+
+class Vec3 {
+public:
+    float x, y, z;
+
+    Vec3() : x(0), y(0), z(0) {}
+
+    Vec3(float x, float y, float z) : x(x), y(y), z(z) {}
+
+    Vec3 operator+(const Vec3& other) const {
+        return Vec3(x + other.x, y + other.y, z + other.z);
+    }
+
+    Vec3 operator-(const Vec3& other) const {
+        return Vec3(x - other.x, y - other.y, z - other.z);
+    }
+
+    Vec3 operator*(float scalar) const {
+        return Vec3(x * scalar, y * scalar, z * scalar);
+    }
+
+    Vec3 operator/(float scalar) const {
+        return Vec3(x / scalar, y / scalar, z / scalar);
```

Comment:
You could use only 1 division by caching the result from 1.f / scalar and multiply by it the components. The compiler could do that instead of us, but it's always good to be on the safe side.